### DISCUSSION TITLE
Logging in Textadept

### DISCUSSION THREAD

**Discussion by eric1234** [orbitalquark/textadept#291] [Show and tell]
I've been customizing my Textadept pretty heavily. Thankfully it is so customizable! But I haven't really figured out a way to debug well.

`ui.statusbar_text` is decent but it only shows the most recent message. `ui.print` can have a list of messages but it opens that up in a buffer which can interfere with what I am trying to debug. Probably any sort of interactive debugging would also only interfere with what I am trying to debug so some sort of logging seems best. Wrote the following quick function to help me do this:

```lua
function(...)
  local args = {...}
  local msg = args[#args]
  if #args == 2 then msg = '['..args[1]..'] '..msg end
  msg = os.date()..' '..msg

  local logger = os.spawn('tee -a '.._USERHOME..'/activity.log')
  logger:write(msg.."\n")
  logger:close()
end
```
I assign this to `textadept.log` then use like this:

```lua
textadept.log('my_module'', 'Some log message')
```

Indicating the module helps filter the log messages but it is optional. All log messages have a date which helps. Spawning a process just to log a message isn't exactly the most efficient but it's the only way I could figure out how to do it via the API I had. Anyway just figured I would share in case it is helpful to anyone else trying to debug their code.

**Comment by orbitalquark** [comment]
That's a good tip. Why not use `io.open(..., 'a+'):write(msg, '\n'):close()`? There shouldn't be any concurrency issues.

For my part, I use `print()` statements (not `ui.print()`, but Lua's `print()`) and run Textadept from a terminal. (I'm on Linux.) Any printed messages are printed to stdout and show up in the terminal.

**Reply by eric1234** [reply]
No reason other than I know nothing about Lua. :)

I'm on Linux for my personal computer but Mac for my work and on Mac I don't think the Lua `print` would work. But good thing to keep in mind for when I am hacking on my personal.

**Reply by rgieseke** [reply]
I think on Mac you can see these in the "Console app".

**Reply by eric1234** [reply]
I gave this a test as if I could get log messages using `print` on the Mac I would probably drop my own impl.

Put a `print('test')` at the start of `init.lua` and no luck of anything showing up in the console app. Did a quick search and looks like it used to behave like that but hasn't [since Mountain Lion](https://stackoverflow.com/a/15655471), mid-2012.

I am thinking of just having my log code (removing the `module` argument) overwrite the stock `print` method when on Mac. That way on Mac I have a way to debug but on Linux I just use the normal stdout.

Thanks for the suggestion though. It was a good thought.
